nomad/core_sched.go

```diff
@@ -706,9 +714,100 @@ func allocGCEligible(a *structs.Allocation, job *structs.Job, gcTime time.Time,
 	return timeDiff > interval.Nanoseconds()
 }
 
-// csiVolumeGC is used to garbage collect CSI volume publications
-func (c *CoreScheduler) csiVolumePublicationGC(eval *structs.Evaluation) error {
-	// TODO: implement me!
-	c.logger.Trace("garbage collecting unclaimed CSI volume publications")
+// csiVolumeClaimGC is used to garbage collect CSI volume claims
+func (c *CoreScheduler) csiVolumeClaimGC(eval *structs.Evaluation) error {
+	c.logger.Trace("garbage collecting unclaimed CSI volume claims")
+
+	// JobID smuggled in with the eval's own JobID
+	var jobID string
+	evalJobID := strings.Split(eval.JobID, ":")
+	if len(evalJobID) != 2 {
+		c.logger.Error("volume gc called without jobID")
+		return nil
+	}
+
+	jobID = evalJobID[1]
+	job, err := c.srv.State().JobByID(nil, eval.Namespace, jobID)
+	if err != nil || job == nil {
+		c.logger.Trace(
+			"cannot find job to perform volume claim GC. it may have been garbage collected",
+			"job", jobID)
+		return nil
+	}
+	c.volumeReap([]*structs.Job{job}, eval.LeaderACL)
+	return nil
+}
+
+// volumeReap contacts the leader and releases volume claims from
+// terminal allocs
+func (c *CoreScheduler) volumeReap(jobs []*structs.Job, leaderACL string) error {
```

Comment:
I might be misreading this - But I don't currently see any validation that no other allocs on the same node are still using the same attached volume (e.g places where the VolumeRequest is equal)

Reply:
Darn, that's a good catch. The `VolumeClaimRequest` we send to release up the claim is scoped to a specific alloc, but the `controllerUnpublishVolume` is scoped to the node.

Effectively there's a "node claim" concept that's emergent from all the alloc claims, but we can derive that here from the `CSIVolume`.